### ISSUE TITLE
long gc pauses when sending large queries using the mysql protocol

### ISSUE THREAD

**Issue by demmer** [youtube/vitess#2849]
We've noticed some long GC pauses in the 1-10ms range when using the mysql protocol.

In the particular use case we are testing, the client opens a mysql connection, sends a large INSERT query (multiple MBs of data) and then closes the connection.

When I test the same frequency of connect/query/disconnect but instead of the large INSERT it does a small SELECT, I don't see the GC pauses. This makes me think it's due to the buffer management code in mysqlconn which ends up doing a bunch of `make([]byte, length)` calls when the packet doesn't fit in the 16K preallocated per-connection buffer.

I'm thinking we can do one of two things to address this:

1. Replace the per-connection `byte[]` buffer with an instance of `bytes.Buffer` and replace the logic in `readEphemeralPacket` / `readPacketDirect` to grow the per-connection buffer to fit the data that needs to be read/written. Then add a `sync.Pool` to wrap the allocation of connection objects so they can be reused across connections that come and go.

2. Keep the existing small per-connection buffers model but replace the various cases where we do `make([]byte, length)` with a `sync.Pool` of temporary `bytes.Buffer` objects that are grown to fit the necessary length. That way we have a shared pool of buffers on hand rather than having to alloc/free them each time.

My inclination is that (1) is a simpler and more straightforward approach, with the potential downside that in the limit we will end up with MAX(connections) * MAX(packet size) memory held in the pool but should at the same time end up being zero-allocation for most calls.

Option (2) should end up being more memory efficient but is more complicated to reason about.

@alainjobart @michael-berlin @sougou: Please let me know what you all think about this.

**Comment by derekperkins** [comment]
We do multi-MB batch inserts on our end using `bytes.Buffer` wrapped in a `sync.Pool` and it works great for us.

```
package pools

import (
	"bytes"
	"sync"
)

var bufferPool = &sync.Pool{
	New: func() interface{} {
		return &bytes.Buffer{}
	},
}

// GetBuffer returns a buffer from the pool.
func GetBuffer() (buf *bytes.Buffer) {
	return bufferPool.Get().(*bytes.Buffer)
}

// PutBuffer returns a buffer to the pool.
// The buffer is reset before it is put back into circulation.
func PutBuffer(buf *bytes.Buffer) {
	buf.Reset()
	bufferPool.Put(buf)
}
```

**Comment by derekperkins** [comment]
@demmer What version of Go did you compile your binaries with? With Go 1.8, you should basically never see GC pauses > 1 ms.

**Comment by michael-berlin** [comment]
We started writing code which only works with Go 1.8 e.g. we use [`sort.Slice()`](https://golang.org/pkg/sort/#Slice) in at least two places? That means @demmer should already be on Go 1.8?

In any case, the `sync.Pool` is a good idea and the way to go for any server implementation. 

@alainjobart can you please comment on the two approaches @demmer suggested?

**Comment by demmer** [comment]
Yes we are running on 1.8. We may switch to 1.8.1 since that was recently released.

**Comment by alainjobart** [comment]
I like @derekperkins suggestion, which is basically suggestion 2 in the original post: whenever we have a memory allocation of a buffer, we replace that with a sync.Pool retrieval of a bytes.Buffer. But the problem I see with it is that we need to return the buffers into the pool after use. So the logic there might be a bit complicated. We might need something similar to the write side, where we have a state and an allocation type, and then the 'send' uses the state to know what to do.

So whenever we read a packet, we remember where it comes from, and we also add a 'free' method that frees the packet memory (by returning its buffer to the pool if it was taken from it). We end up doing a bunch of memory management ourselves, but I guess there is no choice if we want performance here.

Sharing the pool across connections seems like the way to go, to handle rapid short-lived connections better. But if we have the infrastructure to do our own free, we can change the behavor later...

**Comment by sougou** [comment]
I think we should get an estimate of what we'll gain from a change here. What if we ran an experiment with a 5M buffer size instead of 16K?
If that brings down the GC pause, then we'll know it's worth it.

**Comment by alainjobart** [comment]
I have this PR to start the work on this:
https://github.com/youtube/vitess/pull/2874
It mainly refactors the code so we can keep track of our state accurately, and also adds the sync.Pool for the use case of this issue. (similar to @derekperkins suggestion).

@demmer is it easy for you to try this out?

One thing I'd like to do at some point is to see if the pre-allocated buffer is necessary at all, or if we could just use sync.Pool for all buffers.

**Comment by demmer** [comment]
@alainjobart Yes we can give it a try and see how things go later this week.

**Comment by sougou** [comment]
@alainjobart @demmer Here's a recap of my findings so far:

As you can see from the published benchmarks, the parser changes made a big difference, which is awesome.

However, our strategy of giving mysql's ephemeral []byte to ComQuery is risky. There are two failure cases that I could think of (and there may be others):
1. For a scatter stream, we start returning results as soon they start coming. It's theoretically possible that the request might not have been sent to all shards yet at this point. But mysql (package) will release the buffer on the first callback, which will corrupt results that are yet to be sent to other shards.
2. The planbuilder may want to hold on to values (bytes) coming from the parser, which in turn point to the ephemeral buffer. So, they also face the possibility of corruption.

The bottomline is that it will be hard to ensure that no code holds on to the bytes sent to ComQuery.

Fortunately, copying the entire query is cheap. My benchmarks show a negligible difference in the cost if I performed an extra copy of the query (about 0.2ms/MB). So, I'll be changing the mysql package to send a copy  to ComQuery. I'm still debating if it should be a string or []byte. The decision will depend on how this code path intertwines with grpc requests.

**Comment by sougou** [comment]
Updates:

The vttablet benchmarks were flawed because fakesqldb was running expensive regexp matches under the covers. Updated benchmarks:

```
BenchmarkExecuteVarBinary-4          100          14888528 ns/op
BenchmarkExecuteExpression-4        1000           1118654 ns/op
```

The above benchmark is the difference between encoding varbinary bind vars vs. using an already-encoded representation. That's a big difference.

On the VTGate side, using strings in the AST world is an overall win. This is because we need a string representation of the query to put in the query cache, and lots of places convert values to strings. Also, using literal string values further improved performance.

```
VTGate with strings in parser (and literal string values):
BenchmarkWithNormalizer-4            500           3426459 ns/op
VTGate with bytes in parser (no literal string values):
BenchmarkWithNormalizer-4            300           5215343 ns/op
VTGate original:
BenchmarkWithNormalizer-4            100          10927765 ns/op
```

The changes are complex, but the benchmarks clearly show that they're worth it. So, I'll embark on this change. Here's what I hacked up to preview what we can gain: https://github.com/sougou/vitess/commit/aa5f25c23dcb60025673793e9099f89d3c6aba53.
